src/main/java/com/google/vinet/servlets/RegistrationServlet.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.vinet.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.PreparedQuery.TooManyResultsException;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.common.collect.Sets;
+import com.google.gson.Gson;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+@WebServlet("/registration")
+public class RegistrationServlet extends HttpServlet {
+
+  private static final int MIN_NAME_LENGTH = 2;
+  private static final int MAX_NAME_LENGTH = 300;
+  private static final String USER_TABLE_NAME = "UserInfo";
+
+  private enum UserType {
+    ISOLATE,
+    VOLUNTEER
+  }
+
+  private UserService userService;
+
+  public RegistrationServlet() {
+    this.userService = UserServiceFactory.getUserService();
+  }
+
+  public void setUserService(UserService userService) {
+    this.userService = userService;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) {
+    response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+    if (request == null) return;
+
+    if (!userService.isUserLoggedIn()) {
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+      return;
```

Comment:
No logging of the error?

Reply:
I'm not sure I'd count this as an error. Nothing has failed, there simply isn't a user logged in, so the frontend is informed of this. An `SC_UNAUTHORIZED` response implies that the server has *correctly* identified that the process cannot proceed.

Reply:
Oh, this actually wasn't my submission according to `git blame`. Still, I think the design decision is sensible.